src/Type/Php/SprintfFunctionDynamicReturnTypeExtension.php

```diff
@@ -118,6 +152,28 @@ public function getTypeFromFunctionCall(
 		return $this->getConstantType($args, $returnType, $functionReflection, $scope);
 	}
 
+	private function getFormatConstantParts(string $format): ?ConstantStringType
+	{
+		$dummyValues = [];
+		for ($i = 0; $i < self::MAX_INTERPOLATION_RETRIES; $i++) {
```

Comment:
Errr, what is this for?

Reply:
the idea is to replace all placeholders in a pattern with empty string, to detect whether the pattern contains any constant parts and which type of string this constant part has

`%sAA%s` -> `AA` -> non-falsy-string
`%s0%s` -> `0` -> non-empty-string
`0%s0%s` -> `00` -> non-falsy-string

so we know a basic result type for the worst case in which we don't know details about parameters passed to the `sprintf` call:

`sprintf(`%sAA%s`, $s, $s)` -> will always be non-falsy-string, no matter the type of `$s`.

Reply:
Why is this in a for loop with a number of retries if we know the number of placeholders and we know how many values the function call needs? I feel like this could be done completely statically, with some kind of regular expression perhaps.

Reply:
simplified so it no longer loops